compiler/z/codegen/BinaryEvaluator.cpp

```diff
@@ -1898,7 +1949,52 @@ OMR::Z::TreeEvaluator::tryToReplaceLongAndWithRotateInstruction(TR::Node * node,
                // this instruction sets the rotation factor to 0 and sets the zero bit(0x80).
                // So it's effectively zeroing out every bit except the inclusive range of lsBit to msBit
                // The bits in that range are preserved as is
-               generateRIEInstruction(cg, opCode, node, targetReg, sourceReg, msBit, 0x80 + lsBit, 0);
+               int32_t rangeStart = msBit;
+               int32_t rangeEnd = lsBit;
+
+               if (shiftAmount <= 0)
+                  {
+                  if (msBit - shiftAmount > 63)
+                     {
+                     return NULL;
+                     }
+                  else
+                     {
+                     rangeStart = msBit - shiftAmount;
```

Comment:
This looks incorrect. If I understand a `shiftAmount < 0` implies a right shift so we are modifying the selected range by the shift amount? What happens for the following tree:

```
lushr
  land
    x
    lconst 0x4000000000000000
  lconst 1
```

I think `msBit == lsBit == 1` in this case, so `rangeStart = 1 - (-1) == 2` so we'll end up selecting the third most significant bit of 0x4000000000000000 which is 0 and we should get the wrong result here.

Reply:
I don't think your changes should be modifying selection bits, i.e. `rangeStart` and `rangeEnd` should not exist. All your modifications should do is add the rotate amount to the existing `generateRIEInstruction` instead of passing 0, and logic should be added to ensure that the rotate is really a shift.

Reply:
>so we'll end up selecting the third most significant bit of 0x4000000000000000 which is 0.

I think we're actually selecting the third most significant bit of x instead... I have a test case for this specific value, and it does not fail... Unless there's something I'm not getting here?

Reply:
> I don't think your changes should be modifying selection bits, i.e. rangeStart and rangeEnd should not exist. All your modifications should do is add the rotate amount to the existing generateRIEInstruction instead of passing 0, and logic should be added to ensure that the rotate is really a shift.

`rangeStart` and `rangeEnd` merely exist for the "overflow issue (I added some comments on this in my next commit), where the the optimization will still work even if some of the 1's between the msb and lsb are shifted out.

Reply:
> I think we're actually selecting the third most significant bit of x instead

That is exactly the problem. We should be selecting the second most significant bit of (unsigned) _x_ and then shifting right by 1 bit, not selecting the third bit and shifting right by 1. If you replace _x_ with 0x4000000000000000 then the result of that tree when evaluated should be 0x2000000000000000. However I believe your current code will output 0.

Can you generate a log for this case and paste the evaluation of such a tree? I'd like to see the exact `RISBG` instruction that gets generated. I'd also be curious to see the execution, so I'll ask that you hop into gdb and paste the register value before and after the `RISBG` instruction.

Reply:
Ok, apologies I got things mixed up. Just to confirm the rotation happens first, then the bits are selected?

So in the case above we would would rotate _x_ by 63 to the left, meaning a shift right, then we would select the third most significant bit of the _rotated x_ and insert it into the result while zeroing out all the other bits?

Can you generate a log with the tree when _x_ = 0x4000000000000000 and paste the register values before and after the `RISBG` instruction?